docker_swarm: Return UnlockKey

hannseman opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

Return `UnlockKey` when `autolock_managers` is ``true``. It should be returned so it can be retrieved and stored in a safe place to be used if the swarm needs to be restarted.

https://docs.docker.com/engine/swarm/swarm_manager_locking/ 

Returning this parameter could be pretty sensitive if Ansible logs are stored somewhere insecure but I guess the same could be said about the join-tokens.

One alternative could be to create a new module to handle swarm locking / unlocking.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
docker_swarm

hannseman committed:
Return UnlockKey

hannseman committed:
Add changelog fragment

WojciechowskiPiotr commented:
I am not sure if this shouldn't be part of `docker_swarm_info` instead of `docker_swarm`. You can get this key not only when you create the swarm so for me it fits more into `docker_swarm_info`. Or maybe part of both modules?

felixfontein commented:
That's a good question. Having a facility in `docker_swarm_info` which allows to query for the unlocking key is definitely a good idea. But also returning it in `docker_swarm` in certain situations is potentially also good. The only question is whether we want to return it always (as in this PR), or whether we want to add an option which allows to query for it. The second could be more safe for existing users, who might be surprise to suddenly see this key showing up in the logs.

Talking about logs: anyone knows why `signing_ca_key` is not marked as `no_log`?

felixfontein commented:
Also, maybe we should also add a facility which allows to *unlock* swarm managers. I don't think this is currently possible with Ansible. (Thought that should be a separate PR...)

hannseman commented:
> I am not sure if this shouldn't be part of docker_swarm_info instead of docker_swarm. You can get this key not only when you create the swarm so for me it fits more into docker_swarm_info. Or maybe part of both modules?

Ah yes I forgot about `docker_swarm_info`, I've mostly worked on `docker_swarm_service` which does not have an info-module. Totally makes sense to have `docker_swarm_info` return the key.

> But also returning it in docker_swarm in certain situations is potentially also good. The only question is whether we want to return it always (as in this PR), or whether we want to add an option which allows to query for it.

Yes I though about only returning the key when the swarm is initialised but we would also need to return the key when the swarm is updated with `autolock_managers` set to `true`.

One approach might be to just document that the key should be retrieved and stored safely when `autolock_managers` is used and that can be done by using `docker_swarm_info`. 

> Also, maybe we should also add a facility which allows to unlock swarm managers. I don't think this is currently possible with Ansible. (Thought that should be a separate PR...)

Regarding unlocking swarm managers I think it might make sense to create a new module, something like:

```yaml
docker_swarm_lock:
  state: present
  rotate: true
  key: "SWMKEY-1-**"
```

Setting `state` to absent would remove the lock and require that `key` is set.

WojciechowskiPiotr commented:
I think the most optimal solution would be to move the methods to `AnsibleDockerSwarmClient`, implement a feature to query the key in `docker_swarm_info` and return it by default in `docker_swarm` only if Swarm cluster is created. Then add a new module to lock/unlock the cluster. 

I don't see API as a problem here

hannseman committed:
Add method to check if a parameter exists in diffs

hannseman committed:
Add method to get swarm unlock key

hannseman committed:
Add option unlock_key

hannseman committed:
Only return unlock key when created or changed

felixfontein commented on lib/ansible/module_utils/docker/common.py:
```diff
@@ -846,6 +846,14 @@ def get_before_after(self):
             after[item['name']] = item['parameter']
         return before, after
 
+    def exists(self, name):
```
How about calling this `has` instead, or a longer `has_difference_for`?

felixfontein commented on lib/ansible/module_utils/docker/swarm.py:
```diff
@@ -241,3 +244,11 @@ def get_all_nodes_list(self, output='short'):
+        try:
+            return super(AnsibleDockerSwarmClient, self).get_unlock_key()
+        except APIError:
+            return None
```
Suppressing errors is usually not a good idea, especially not in library-level code. If this is really necessary, it should be done where this code is called.

hannseman replied:
I fully agree. Had some thought about that it would return 404 on unlocked swarms but it does return an empty string along with a HTTP 200 instead. I've added a test to verify this. In `docker_swarm` I think we should catch `APIError` to not fail the entire task just because the key retrieval failed.

hannseman commented:
I've modified `docker_swarm` to only check and return the unlock key when the swarm is initialised or when the swarm is updated with `autolock_managers` set as `true`.

 I've also added the option `unlock_key` to `docker_swarm_info` which will return the key if applicable. 

I'll update the change log after we've agreed on the option naming.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm.py:
```diff
@@ -230,6 +230,12 @@
                   example: SWMTKN-1--xxxxx
+      UnlockKey:
+          description: The swarm unlock-key if I(autolock_managers) is C(true).
+          returned: success
```
Shouldn't this be something like `returned: on success if I(autolock_managers) is C(true)`? Also, I thought you only want to return it if the swarm was actually created? Maybe it is better to relax that condition, and always return it when the module is called with `state == present` and `autolock_managers == true`. Then the rest of the playbook doesn't has to know whether the swarm was actually created or not.

hannseman replied:
It does not make sense to only return it when the swarm is created. If `autolock_managers` is set to `true` by an update one would expect the key to be returned. This is also the way the docker cli works. In other words if the swarm is created without a lock and later updated with autolock the key is returned. If the update is repeated with autolock the key is not returned.

Returning the key only on changes to `autolock_managers` will reduce the runtime of the task as it has to make one less API request. I'll also make it a tiny bit safer as the key is not always returned and logged.

Although I agree on the more verbose return description.

WojciechowskiPiotr replied:
It does make sense to return when the swarm is created or when the key is changed. In any other case, it can be read by `docker_swarm_info`. Actually, I think the plan was to remove all `inspect` features from `docker_swarm`. @felixfontein do we have time to do that?

felixfontein replied:
I'm mainly concerned that the condition when that value is returned might be too confusing for users who want to use it. But in the end I'm fine when it's properly documented when exactly it is returned, and when a pointer is added to `docker_swarm_info` so that people know they can get it reliably from there.

@WojciechowskiPiotr we can deprecate `inspect` (in fact we already did that), but it has to stay until Ansible 2.12.

hannseman replied:
I added documentation about that `docker_swarm_info` can be used to retrieve the unlock key under the `autolock_managers` option.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_info.py:
```diff
@@ -119,6 +124,11 @@
+    unlock_key: yes
+  register: result
+
 - debug:
```
I think you should move this `debug` up before the example, and add a new `debug` here which outputs `result.swarm_unlock_key`.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm.py:
```diff
@@ -427,11 +436,26 @@ def inspect_swarm(self):
+        default = {'UnlockKey': None}
+        if not self.has_swarm_lock_changed():
+            return default
+        return self.client.get_unlock_key() or default
```
Does `self.client.get_unlock_key()` returns something of form `{ 'UnlockKey': xxx }` (as this function seems to assume), or does it just return `xxx` (as the function in `docker_swarm_info` seems to assume)?

hannseman replied:
It does return `{ 'UnlockKey': xxx }`. The method in `docker_swarm_info` actually unpacks that dictionary:

```python
def get_docker_swarm_unlock_key(self):
    unlock_key = self.client.get_unlock_key()
    return unlock_key.get('UnlockKey')
```

In my opinion it made a cleaner API on `docker_swarm_info` to just return the key in the variable `swarm_unlock_key` while it made more sense in `docker_swarm` to return `UnlockKey` along with the other tokens (`JoinTokens`) in the same dictionary .

felixfontein commented on test/integration/targets/docker_swarm_info/tasks/test_swarm_info.yml:
```diff
@@ -125,6 +125,26 @@
+  - name: assert reading swarm facts and swarm unlock key
+    assert:
+      that:
+        - 'output.swarm_unlock_key'
```
I would make this a little bit more explicit:
```suggestion
        - 'output.swarm_unlock_key is string'
```

felixfontein commented:
While you're at it: can you change `@wojciechowskipiotr` in the author section of `docker_swarm_info` to `@WojciechowskiPiotr`? This difference in casing caused the extra CC for @WojciechowskiPiotr.

hannseman committed:
Rename difference check

hannseman committed:
Extend unlock key example

hannseman committed:
Assert that unlock_key is a string

hannseman committed:
Fix docker_swarm_info authors

hannseman committed:
Don’t silence APIErrors

hannseman committed:
Test unlock_key on unlocked swarm

hannseman committed:
Catch APIError when retrieving unlock key

hannseman committed:
Better return value description

hannseman committed:
Lint

WojciechowskiPiotr commented on lib/ansible/modules/cloud/docker/docker_swarm.py:
```diff
@@ -427,11 +437,29 @@ def inspect_swarm(self):
         except APIError:
             return
 
+    def get_unlock_key(self):
```
I think this should be in `AnsibleDockerSwarmClient` class

felixfontein replied:
I disagree. This is only needed for `docker_swarm`, not for `docker_swarm_info`.

hannseman replied:
I too disagree. It's very `docker_swarm` specific as it tests if the module-parameter `autolock_managers` has changed. The default handling of returning the raw `UnlockKey` key can be argued to be lifted into `AnsibleDockerSwarmClient` as discussed in https://github.com/ansible/ansible/pull/54490#discussion_r270600286. But I do think we need a specific method here to do the diff-checking.

WojciechowskiPiotr commented on lib/ansible/modules/cloud/docker/docker_swarm.py:
```diff
@@ -427,11 +437,29 @@ def inspect_swarm(self):
+        except APIError:
+            return default
+
+    def has_swarm_lock_changed(self):
```
This as well

felixfontein replied:
This *cannot* be in `DockerSwarmClient` since it does something very module-specific.

WojciechowskiPiotr commented on lib/ansible/modules/cloud/docker/docker_swarm_info.py:
```diff
@@ -305,6 +325,10 @@ def get_essential_facts_services(item):
 
         return object_essentials
 
+    def get_docker_swarm_unlock_key(self):
```
This should be in `AnsibleDockerSwarmClient` class as well in my opinion and either unified for both `docker_swarm` and `docker_swarm_info` maybe?

felixfontein replied:
Please check https://github.com/ansible/ansible/pull/54490#discussion_r270585152
I think it makes more sense the way it is now: `docker_swarm` needs the structure with `UnlockKey` in it, while `docker_swarm_info` needs the raw value.f

WojciechowskiPiotr replied:
But it may be a universal method in the `AnsibleDockerSwarmClient` class where you add a parameter to define if you want raw value or structure. The process of reading key is shared between two modules that why it fits more into a shared library in my opinion. And in case a new module for locking/unlocking needs it as well it will add the third module to use it.

felixfontein replied:
I personally don't like functions with variable return type (depending on an argument). It's just impossible to verify code using such functions with static code analysis.

I would not start adding methods to module_utils/docker/swarm.py just because we might need them in the future, but do that refactoring when we actually add a new module that needs them. Then we'll find out what exactly we need.

WojciechowskiPiotr replied:
In my proposal, it is a method shared already between two modules. But of course, we can move and consolidate later

hannseman replied:
My proposal is to wait until we implement the locking-module and see what kind of implementation is to be used there (dict with `UnlockKey` or just a string). The common implementation can then be lifted to `AnsibleDockerSwarmClient` and the other one can be used as an override.

WojciechowskiPiotr replied:
Ok, sounds good

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm.py:
```diff
@@ -230,6 +230,13 @@
+      UnlockKey:
+          description: The swarm unlock-key if I(autolock_managers) is C(true).
+          returned: on success if I(autolock_managers) is C(true)
+            and swarm is initialised or autolock_managers is changed.
```
```suggestion
            and swarm is initialised, or if I(autolock_managers) has changed.
```

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_info.py:
```diff
@@ -305,6 +325,10 @@ def get_essential_facts_services(item):
         return object_essentials
 
+    def get_docker_swarm_unlock_key(self):
+        unlock_key = self.client.get_unlock_key()
```
```suggestion
        unlock_key = self.client.get_unlock_key() or {}
```
Just to avoid that the next line fails.

felixfontein commented on test/integration/targets/docker_swarm_info/tasks/test_swarm_info.yml:
```diff
@@ -125,6 +125,39 @@
+  - name: assert reading swarm facts and non existing swarm unlock key
+    assert:
+      that:
+        - 'not output.swarm_unlock_key'
```
```suggestion
        - 'output.swarm_unlock_key is none'
```

felixfontein replied:
And maybe in some of the tests above: `output.swarm_unlock_key is not defined`

hannseman replied:
The returned value from the Docker API actually is an empty string, hence the `not` assertion. But maybe it makes a cleaner API to set `swarm_unlock_key` to `None` if it's an empty string.

hannseman replied:
Something like:

```python
    def get_docker_swarm_unlock_key(self):
        unlock_key = self.client.get_unlock_key()
        return unlock_key.get('UnlockKey') or None
```

felixfontein committed:
Fix UnlockKey return value documentation

Co-Authored-By: hannseman <hannes@5monkeys.se>

felixfontein committed:
Get unlock key safely

Co-Authored-By: hannseman <hannes@5monkeys.se>

hannseman committed:
Return None on empty UnlockKey

hannseman committed:
Assert swarm_unlock_key is undefined if unqueried

hannseman committed:
Add documentation about swarm_info unlock_key

hannseman committed:
Add change log fragment for unlock_key option

felixfontein commented on changelogs/fragments/54490-docker_swarm-return-unlock-key.yaml:
```diff
@@ -1,2 +1,3 @@
 minor_changes:
   - "docker_swarm - ``UnlockKey`` will now be returned when ``autolock_managers`` is ``true``."
+  - "docker_swarm_info - Add option ``unlock_key`` to retrieve the swarm unlock key."
```
`docker_swarm_info` has been added for Ansible 2.8, so it does not need a changelog fragment.

hannseman replied:
I missed that. Fixed.

felixfontein commented on lib/ansible/module_utils/docker/common.py:
```diff
@@ -846,6 +846,14 @@ def get_before_after(self):
+        '''
+        Returns a boolean if a difference exists for name
+        '''
+        return any([
```
How about removing the `[ ]`? That should be more efficient, since `any()` can iterate over the generated sequence and stop when it finds the first match. With the array, first the whole list is generated, and then `any()` iterates over the list elements until it finds a match.

hannseman replied:
Good catch!

hannseman committed:
Revert "Add change log fragment for unlock_key option"

This reverts commit e3cb2325b552e5d14cc3f42b33a86bf3ee84d3b9.

hannseman committed:
Use generator expression instead

hannseman commented:
Hmm the tests have failed by getting stuck on `docker_swarm : Cleanup (trying)` a couple of times (over 45 minutes limit).

https://app.shippable.com/github/ansible/ansible/runs/116906/59/console

felixfontein commented:
Hmm, yes, I've already seen that once yesterday. This is really strange. The `docker_swarm` tests are known for being problematic; docker daemon apparently doesn't like it if it is reconfigured too often. Let's try something else, like killing it and removing `/var/lib/docker`, and then starting it again.

Can you try to replace the
```.yaml
  - name: Restart docker daemon
    service:
      name: docker
      state: restarted
    become: yes
```
block by
```
  - name: Kill docker daemon
    command: systemctl stop -s 9 docker
    become: yes
  - name: Restart docker daemon
    service:
      name: docker
      state: started
    become: yes
```
and see if that helps?

hannseman committed:
Restart docker more decisively

hannseman commented:
Looks like it didn’t help. Maybe clearing out `/var/lib/docker` might work.

felixfontein commented on test/integration/targets/docker_swarm/tasks/main.yml:
```diff
@@ -13,11 +13,16 @@
     ignore_errors: yes
 
+  - name: Kill docker daemon
+    command: systemctl stop -s 9 docker
```
Ah, I think I misread the `systemctl` documentation; I think signals can only be specified for `systemctl kill`. Can you try the following?
```suggestion
    command: systemctl kill -s 9 docker
```

felixfontein committed:
Use systemctl kill

Co-Authored-By: hannseman <hannes@5monkeys.se>

felixfontein commented on test/integration/targets/docker_swarm/tasks/main.yml:
```diff
@@ -13,11 +13,16 @@
     service:
       name: docker
-      state: restarted
+      state: started
```
One last try: can you change this back to `restarted`? I somehow have the feeling that `systemctl kill` doesn't put the unit into `stopped` state, so `state: started` doesn't do anything.

hannseman replied:
Done. Maybe I'm missing something but a lot of failures have been in the task before these (named Cleanup (trying)). This is where some CI timeouts have occurred. Maybe one option could be to use `command` with `timeout 1m docker swarm leave -f`?

felixfontein replied:
Interesting. I wasn't aware of that task taking forever, I thought it fails a lot (that's why it got `ignore_errors`), but I wasn't aware that it hits the 45 min timeout. Do you know where that happened?

From the runs I found in this PR (https://app.shippable.com/github/ansible/ansible/runs/116966/59/console, https://app.shippable.com/github/ansible/ansible/runs/116906/59/console, https://app.shippable.com/github/ansible/ansible/runs/116901/59/console, https://app.shippable.com/github/ansible/ansible/runs/116898/59/console, https://app.shippable.com/github/ansible/ansible/runs/116838/58/console, https://app.shippable.com/github/ansible/ansible/runs/116835/59/console), the problem was always the `Restart docker daemon` task.

hannseman replied:
Yes you're right. I missed that CI continued on to run `Restart docker daemon` after it logged an error message in `Cleanup (trying)`.

ffdybuster replied:
Can you trigger another CI run, so we can see whether the new cleanup works more reliably?

felixfontein replied:
Can you try another time? Just so it hopefully works three times in a row ;)

felixfontein commented:
> Looks like it didn’t help. Maybe clearing out `/var/lib/docker` is an option.

I would keep that as the last resort, since it's really destructive. In particular, I don't want to run that on my own machine :)

hannseman commented:
> In particular, I don't want to run that on my own machine :)

Haha that's true.

hannseman committed:
Try to restart docker daemon

felixfontein commented:
Tests passed three times in a row, so I guess this problem is fixed. Or at least got seldom enough :-)

shipit

hannseman commented:
@felixfontein @WojciechowskiPiotr thank you for the review and @gundalow thank you for merging!

felixfontein commented:
@hannseman thanks for implementing this!
